### PR TITLE
Add support for spatial audio

### PR DESCRIPTION
**Pull request by GitGhillie**
See the added example spatial.rs.

Adds components to handle FMOD events in a Bevy way, by implementing AudioSinkPlayback and storing a reference to the FMOD event. The code is pretty rough but it works. In your FMOD project you just need to have a spatializer on your event, which is added by default in FMOD Studio.

### CODE REVIEW

File: src/plugin.rs
Comment (Salzian) [inline src/plugin.rs:275]: Having `// TODO`'s is ok, but we should keep the main branch clean of unimplemented API.

File: src/plugin.rs
Comment (Salzian) [inline src/plugin.rs:34]: This is basically `PlaySoundEvent`. We should decide on one approach before merging this.

**Review by Salzian — CHANGES_REQUESTED** [review]
This is a surface level review. I have to dig a bit deeper into the API design implications.

File: src/plugin.rs
Comment (Salzian) [reply src/plugin.rs:43]: Created a PR in libfmod-gen so we may not have to do this in the future anymore.

https://github.com/lebedec/libfmod-gen/pull/10

File: src/plugin.rs
Comment (Salzian) [inline src/plugin.rs:192]: This TODO should be an issue.

Also, by using triple slash `///` you expose this to-do comment to the user. I don't think that makes sense in this case.

File: src/plugin.rs
Comment (Salzian) [inline src/plugin.rs:202]: This TODO should be an issue.

File: src/plugin.rs
Comment (GitGhillie) [reply src/plugin.rs:29]: If we do this then we expose `previous_position` to the user which seems a bit odd to me. I'd rather the user does not have to worry about transforms when initializing the listener. Let me know what you think

File: examples/spatial.rs
Comment (GitGhillie) [reply examples/spatial.rs]: Maybe it's for the doc scraping they do for the examples? Adding just in case

File: src/plugin.rs
Comment (GitGhillie) [reply src/plugin.rs:40]: `UnsafeEventInstance` should not be pub imo because it's something internal to the plugin that the user shouldn't have to interact with if we make the API nice enough. Some time after I placed this comment I implement Drop so it doesn't need to be pub anymore, just forgot to address the comment

File: src/plugin.rs
Comment (GitGhillie) [reply src/plugin.rs]: Implemented the closest thing I could find (increasing the pitch speeds the audio up...)

File: src/plugin.rs
Comment (Salzian) [reply src/plugin.rs]: How does this expose `previous_position`?

Comment (GitGhillie) [reply src/plugin.rs]: When you remove the `Default` impl the user will have to initialize `previous_position`. Can't add a component without initializing all fields

File: src/plugin.rs
Comment (GitGhillie) [reply src/plugin.rs]: Latest commit contains my take on `PlaySoundEvent` which doesn't use Bevy events. However, it's still a different approach to what I did for spatial audio. I think if we go for one approach either the minimal example will become a bit more complicated (i.e. spawn an entity which plays an event once and then despawns itself) or my implementation for spatial will become a lot different and further removed from what I think is idiomatic. Let me know what you think

Comment (GitGhillie) [reply src/plugin.rs]: Btw I'm not saying my take on `PlaySoundEvent` is the way to go, just spitballing

File: src/play_sound_event.rs
Comment (Salzian) [inline src/play_sound_event.rs:13]: Haven't discussed this yet, but I'd like to avoid the use of `unwrap`, especially since this project is a library. 

I prefer to use the `anyhow` crate with `bevy_mod_sysfail` whenever possible. In some cases, you'll have to resort to `expect`. It's always valuable for developers to know what went wrong with a descriptive error message.

File: src/play_sound_event.rs
Comment (Salzian) [inline src/play_sound_event.rs:4]: There's already a `NonSend<Studio>` resource. Maybe try moving this logic into a system?

File: src/utils.rs
Comment (Salzian) [inline src/utils.rs:1]: Utils folders are a code smell 👃 Can you come up with a better name for this module?

https://mykeels.medium.com/utils-are-a-blackhole-c8320968325b

This function is partially necessary to convert between coordinate systems. Maybe we can quick-fix #21 first?

File: src/sync_wrapped.rs
Comment (Salzian) [inline src/sync_wrapped.rs:4]: nit: When this file grows, I think it makes more sense to

1. swap the train impls with the struct definition
2. move EventInstance into `src/sync_wrapped/event_instance.rs

File: src/audio_player.rs
Comment (Salzian) [inline src/audio_player.rs:1]: This file still contains multiple structs. I know rustaceans love dumping everything into one file. I prefer separate files for easier navigation. What do you think? 🙂

File: src/audio_player.rs
Comment (Salzian) [inline src/audio_player.rs:7]: This could be a `Bundle` using `Transform`. That way we can implement [direction and elevation](https://www.fmod.com/docs/2.02/studio/parameters-reference.html#direction) in the future.

File: src/audio_player.rs
Comment (Salzian) [inline src/audio_player.rs:6]: What exactly does **previous**_position mean? Is it delayed by a frame?

File: src/audio_player.rs
Comment (Salzian) [inline src/audio_player.rs:24]: This is unused.

File: src/audio_player.rs
Comment (Salzian) [inline src/audio_player.rs:26]: Could be a `Bundle` with `Transform` again.

File: src/audio_player.rs
Comment (Salzian) [inline src/audio_player.rs:18]: Just thinking out loud: Can AudioSource and AudioSourcePlayer be merged into one struct? I don't see much reason in separating them right now.

File: src/audio_player.rs
Comment (Salzian) [inline src/audio_player.rs:39]: Self-explanatory 😉

File: src/play_sound_event.rs
Comment (Salzian) [inline src/play_sound_event.rs:1]: File naming is outdated.

File: src/audio_player.rs
Comment (GitGhillie) [reply src/audio_player.rs:6]: Yes, it's used to calculate the velocity for the 3dAttributes

File: src/audio_player.rs
Comment (GitGhillie) [reply src/audio_player.rs:7]: Direction and elevation are part of the built-in parameters, they will get updated automatically based on the 3dAttributes we set. For something like directivity we either have to use user parameters (aka local parameters) or use 3rd party plugins afaik

File: src/audio_player.rs
Comment (GitGhillie) [reply src/audio_player.rs:18]: Let's say it's combined into AudioSource, then the moment the user spawns this the EventInstance will need to be initialized, which means you need a reference to the Studio. I tried but I'm not sure how to do it without making the example more complicated

File: src/audio_player.rs
Comment (GitGhillie) [reply src/audio_player.rs:26]: By the way, if we bundle it with Transform then with the current implementation the Attributes3d will be calculated every frame even for 2D sounds and menu music. Either way we should probably add an example for 2D audio soon so it will also be clear for ourselves what that looks like

File: src/audio_player.rs
Comment (GitGhillie) [reply src/audio_player.rs:1]: Agree actually, personally I see removing unwraps and splitting into separate files as a finalizing step, but especially the latter I should do earlier to make it easier to review :)

File: src/play_sound_event.rs
Comment (GitGhillie) [reply src/play_sound_event.rs:4]: Hmm I'm not sure how we would do that. Unless we make the user spawn an entity with a OneShotPlayer component for example

File: src/play_sound_event.rs
Comment (GitGhillie) [reply src/play_sound_event.rs:13]: I couldn't get it to work with sysfail here so I went with `expect`. Other unwraps have been removed as well

File: src/utils.rs
Comment (GitGhillie) [reply src/utils.rs:1]: To be honest I don't know how to fix #21, that flag is part of FMOD core initialization while we init FMOD studio. You may be able to init the core but I believe you need to init FMOD studio to enable Live Update as well.

File: src/plugin.rs
Comment (Salzian) [reply src/plugin.rs]: Clarified on Discord. Will implement...

**Comment by Salzian** [comment]
Moved to #26

**Pull request closed without merging**
